concent_api/core/message_handlers.py

```diff
@@ -79,7 +79,7 @@ def handle_send_force_report_computed_task(client_message):
         task_to_compute,
     )
 
-    if Subtask.objects.filter(  # pylint: disable=no-member
+    if Subtask.objects.select_for_update().filter(  # pylint: disable=no-member
         subtask_id=task_to_compute.compute_task_def['subtask_id'],
     ).exists():
```

Comment:
This is not wrong but not really necessary. If it exists, there's not need to lock it because we exit immediately. If it does not, you won't be able to lock it anyway.

Reply:
Maybe it would be better to change it to `try/except DoesNotExists` flow.

Reply:
I'd personally just get the object first with `filter()` and then check the count. The `try ... except` version is very verbose. I'd leave exceptions for cases where they're really needed.

Reply:
So it would be still two queries, which makes version with `exists` better. Doing this with `try/except DoesNotExists` is a typical Django flow.

Reply:
Yes, but as @cameel wrote in first comment- it is not necessary. If subtask exist flow ends with duplicate request. If not it will not be locked anyway. I simply deleted `select_for_update` in this place ane left it as at the beginning

Reply:
> Doing this with try/except `DoesNotExists` is a typical Django flow.

Honestly, I often see people arguing for either as more typical or better. They're not always equivalent but when they are I think it's mostly a personal preference. I prefer not using exceptions in this particular case but I don't want to force my preference here so I'll just explain why I prefer it: the `try ... except` really stands out to me as something "heavy" and more complex. A version with `exists()` seems plain and ordinary. When you start catching stuff it looks like you're handing an exceptional case and such cases often need closer review. If it's a simple existence check with an `if`, I know that there isn't much that can go wrong and I can simply move on without focusing on it.

It's not wrong to do it with an exception but it just seems like an overkill to me.

Reply:
Right, but it is still two queries to database against one, which in applications which use database a lot can have serious impact (if those double queries vs one is repeated several times). We can always switch to:
```
subtask = Subtask.objects.filter(subtask_id=value).first()
if subtask is not None:
    do_stuff()
```

Which is single query and doesn't involve exceptions BUT doesn't raise exception when more than 1 object is returned (here `subtask_id` is unique so it won't happen, but its easy to forget that)

Actually the `try/except` is suggested in official Django documentation.

Reply:
That's why I never use `first()`. I'd write it as:

``` python
subtasks = Subtask.objects.filter(subtask_id=value)
assert len(subtasks) <= 1
if len(subtasks) == 1:
    do_stuff()
```

It's just one query and you also get the check correctly as an `assert` (because it's not necessary once the code has been debugged).

I usually create helper like this:
``` python
def get_one_or_none(query):
    assert len(query) <= 1
    if len(query) == 0:
        return None
    return query[0]
```

or:
``` python
def get_one_or_none(model, filter_condition):
    query = model.objects.filter(filter_condition)
    assert len(query) <= 1
    if len(query) == 0:
        return None
    return query[0]
```

which reduces the code to:
``` python
if get_one_or_none(Subtask, subtask_id=value) is not None:
    do_stuff()
```

and is nicer and more concise than:
``` python
try:
    subtask = Subtask.objects.get(subtask_id=value)
except Subtask.DoesNotExist:
    pass
else:
    do_stuff()
```

You could put the version with an exception into a function too - and maybe that would be the best compromise because it would still use exceptions but would hide the `try ... except` block.